server/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java

```diff
@@ -109,26 +111,24 @@ protected NodeGatewayStartedShards nodeOperation(NodeRequest request, Task task)
             ShardStateMetaData shardStateMetaData = ShardStateMetaData.FORMAT.loadLatestState(logger, namedXContentRegistry,
                 nodeEnv.availableShardPaths(request.shardId));
             if (shardStateMetaData != null) {
-                IndexMetaData metaData = clusterService.state().metaData().index(shardId.getIndex());
-                if (metaData == null) {
-                    // we may send this requests while processing the cluster state that recovered the index
-                    // sometimes the request comes in before the local node processed that cluster state
-                    // in such cases we can load it from disk
-                    metaData = IndexMetaData.FORMAT.loadLatestState(logger, namedXContentRegistry,
-                        nodeEnv.indexPaths(shardId.getIndex()));
-                }
-                if (metaData == null) {
-                    ElasticsearchException e = new ElasticsearchException("failed to find local IndexMetaData");
-                    e.setShard(request.shardId);
-                    throw e;
-                }
-
                 if (indicesService.getShardOrNull(shardId) == null) {
+                    final String customDataPath;
+                    if (request.getCustomDataPath() != null) {
+                        customDataPath = request.getCustomDataPath();
+                    } else {
+                        // TODO: Fallback for BWC with older ES versions. Remove once request.getCustomDataPath() always returns non-null
```

Comment:
Could this be an `assert` to make sure we clean it up in 9.0?

Reply:
I wanted to clean this up earlier, namely in 8.0 once we have 7.6 released. We only need to consider wire compatibility here. I don't want to insert a timebomb for a minor here though as it makes version bumping unnecessarily difficult.